### ISSUE TITLE
unexpected search command breaks imap search

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1977]
|<img alt="tvachon's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| tvachon reported|
|-|-|
|Trac ID|trac#1977|
|Type|defect|
|Created|2006-08-01 00:05:54Z|
|Branch|https://github.com/twisted/twisted/tree/imap4-search-1977-2|

twisted.mail.imap4.IMAP4Server.singleSearchStep has code to handle unexpected search commands, but this code cannot be reached (line 1466):
```
            f = getattr(self, 'search_' + c)
            if f:
                if not f(query, id, msg):
                    return False
            else:
                # IMAP goes *out of its way* to be complex
                # Sequence sets to search should be specified
                # with a command, like EVERYTHING ELSE.
                try:
                    m = parseIdList(c)
                except:
                    log.err('Unknown search term: ' + c)
                else:
                    if id not in m:
                        return False
```
Since f = getattr(self, 'search_' + c) raises an AttributeError if 'search_' + c does not exist, f can never evaluate to False, so the "else" suite never runs. 

The fix is to replace 
```
f = getattr(self, 'search_' + c) 
```
with 
```
f = getattr(self, 'search_' + c, None)
```

This problem arises from IMAP client commands like
```
00000003 SEARCH 1 UNDELETED UNSEEN
```

The only client I've seen use this is Pine, but since this does break compatibility with Pine, I've marked its priority as High.

Attachments:

* [ticket1977fix_against20060731svn.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/15a/15afd7262ebd18f6f7c1401f13249c2afd747402/8a53f7a41a12d0d69a78fee63429c9ff669cae0d.diff) (412 bytes) - added by tvachon on 2006-08-01 00:37:35Z - fix for this defect
* [ticket1977_unittest.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/15a/15afd7262ebd18f6f7c1401f13249c2afd747402/953ea8fff11b334fe15827c4d115392c09d472fd.diff) (1474 bytes) - added by tvachon on 2006-08-14 23:05:10Z - Unittest for this defect

<details><summary>Searchable metadata</summary>

```
trac-id__1977 1977
type__defect defect
reporter__tvachon tvachon
priority__high high
milestone__ 
branch__branches_imap4_search_1977_2 branches/imap4-search-1977-2
branch_author__therve__itamar__exarkun therve, itamar, exarkun
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__imap4 imap4
time__1154390754000000 1154390754000000
changetime__1258294834000000 1258294834000000
version__None None
owner__ 
cc__exarkun cc__tvachon
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="tvachon's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|tvachon commented|
|-|-|

This and ticket [#1978](https://github.com/twisted/twisted/issues/1978) are my first bug submissions to Twisted. I expect that just submission of these patches is not enough to get them included, and would be more than happy to follow up with whatever steps are needed to move the inclusion process along.

Thanks,

Travis Vachon

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

These fixes will probably be included in due course.  The thing you can do which will move them along the fastest is to write unit tests.  No new code (even bugfixes!) goes into Twisted without tests: otherwise we don't have a rigorous way of determining that the bug was actually fixed, fixed on all platforms, remains fixed, etc.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to tvachon|
|-|-|

Thanks for the patch and test.  I modified the test a bit to actually assert something about the results from search().  Unfortunately, search support with message sets seems to be completely broken.  The patch prevents an exception from being raised, but does not correctly restrict the result set.

Searching for 1 should return at most a result set containing 1.  However, with Twisted's search implementation, it returns the entire contents of the mailbox.  I'm not exactly sure why this is right now.

I've checked your patch and my modifications into a branch, [branches/imap4-search-1977,](branches/imap4-search-1977,) if you'd like to take a look and make further changes.

I probably won't get back to this for at least a week, quite possibly longer.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

btw, that link won't work until tomorrow.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

I can't see any reason this needs to be targeted at 9.0 (8.2+1).

**Comment by twisted-trac** [comment]
|<img alt="jpsimons's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|jpsimons commented|
|-|-|

I ran into the same issue. The iPhone's IMAP client sends a search like `1:* NOT DELETED` which ends up hitting this bug. The problem is you can fix it like you mention with a None default for get_attr, but then it's running new code than never ran before, which is itself buggy. My `1:*` creates a MessageSet with an undefined `last` property, so __contains__ ends up failing with the *. I patched it like this (there's a better way I'm sure):

```
            c = q.upper()
-           f = getattr(self, 'search_' + c)
+           f = getattr(self, 'search_' + c, None)
            if f:
                if not f(query, id, msg):
                    return False
            else:
                # IMAP goes *out of its way* to be complex
                # Sequence sets to search should be specified
                # with a command, like EVERYTHING ELSE.
                try:
                    m = parseIdList(c)
+                   m.last = 99999999999
                except:
                    log.err('Unknown search term: ' + c)
                else:
                    if id not in m:
                        return False
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

[#4048](https://github.com/twisted/twisted/issues/4048) was a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve commented|
|-|-|

(In [27367]) Branching to 'imap4-search-1977-2'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve commented|
|-|-|

I went on and had a look. The branch wasn't merging cleanly, and the test didn't pass anymore. I fixed it, and fix the other problem with '1:*'. I can't review that anymore, though.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

There are some IMAP4 issues. :/  Itamar and I will try to fix them.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Okay!  Hopefully things are a bit better now.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve set owner to @exarkun|
|-|-|

Alright! Following the great rule that I reviewed what I didn't write, please merge this one.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set status to `closed`|
|-|-|

(In [27483]) Merge imap4-search-1977-2

Author: tvachon, therve, exarkun, itamar
Reviewer: exarkun, itamar, therve
Fixes: [#1977](https://github.com/twisted/twisted/issues/1977)

Fix `twisted.mail.imap4.IMAP4Server`'s handling of the SEARCH command so that
it handles the bare `<sequence set>` case.  Previously this would result in an
unhandled exception.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set status to `reopened`|
|-|-|

(In [27489]) Revert r27483 - test suite regression on Python < 2.6

Reopens [#1977](https://github.com/twisted/twisted/issues/1977)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

(In [27490]) Use zip range len + 1 instead of enumerate because Python 2.5 and earlier don't have two-argument enumerate

refs [#1977](https://github.com/twisted/twisted/issues/1977)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun set status to `closed`|
|-|-|

(In [27499]) Merge imap4-search-1977-2

Author: tvachon, therve, exarkun, itamar
Reviewer: exarkun, itamar, therve
Fixes: [#1977](https://github.com/twisted/twisted/issues/1977)

Fix `twisted.mail.imap4.IMAP4Server`'s handling of the SEARCH command so that
it handles the bare `<sequence set>` case.  Previously this would result in an
unhandled exception.

This remerge replaces the use of a Python 2.6-only construct (two-arg enumerate)
with a more Python 2.3-friendly construct.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
